refactor: allow setting a db user name different from the database name

akhilnarang opened the pull request:
This will allow us to use a separate database user, and not only one matching the database name

Resolves #23323

akhilnarang committed:
refactor: allow setting a db user name different from the database name

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

gavindsouza commented on frappe/integrations/offsite_backup_utils.py:
```diff
@@ -110,7 +110,7 @@ def generate_files_backup():
 	backup = BackupGenerator(
 		frappe.conf.db_name,
-		frappe.conf.db_name,
+		frappe.conf.db_user or frappe.conf.db_name,
```
I think it's best to use `frappe.conf.db_user` everywhere and make sure it's always set instead of defining these fallbacks everywhere (like `frappe.conf.db_user ||= frappe.conf.db_name`).

ankush replied:
All existing site configs will have to be updated for this to work. :eyes:

ankush replied:
We can write a patch but we can't do patch + removing fallback together AFAIK. Maybe we should write a patch for now and remove fallbacks from next version.

gavindsouza replied:
I meant during `frappe.init`, not in the site config

akhilnarang replied:
Yep makes sense, something like 329fcf9fe8f0ca2838e9d8897a6751ba1ab03d0b?

ankush commented on frappe/app.py:
```diff
@@ -171,6 +171,10 @@ def init_request(request):
+	# Set the user as database name if not set in config
+	if frappe.local.conf.db_user is None:
+		frappe.local.conf.db_user = frappe.local.conf.db_name
+
```
This should be done in `init` or wherever conf is init-ted. `app.py` only applies to requests.

ankush commented on frappe/utils/backups.py:
```diff
@@ -485,7 +485,7 @@ def fetch_latest_backups(partial=False):
 	odb = BackupGenerator(
 		frappe.conf.db_name,
-		frappe.conf.db_name,
+		frappe.conf.db_user or frappe.conf.db_name,
```
This shouldn't be required if conf is updated correctly.

akhilnarang committed:
fix(setup_db): use the correct user

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: set a fallback for db user in `connect()`

In case we don't have a site name, `init()` won't be called, and we won't have a username

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor(replica): allow using separate db user name

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
refactor: add in a fallback to use db name as db user name

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: add in missing logic, add in a test for creating a site with a specific user

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

gavindsouza reviewed (commented):
Interesting CI failure:

```py
psycopg2.errors.SyntaxError: syntax error at or near "830"
LINE 1: CREATE USER 830c1ee761f5cbf75bc547a7dea2137a66a6546d359a5a77...
                    ^
```

Pg prefers roles that start with alpha-characters instead. I don't know if some change in this PR changed some existing behaviour (I didn't dig through), but can this be taken care of, too?

akhilnarang commented:
> Interesting CI failure:
> 
> ```python
> psycopg2.errors.SyntaxError: syntax error at or near "830"
> LINE 1: CREATE USER 830c1ee761f5cbf75bc547a7dea2137a66a6546d359a5a77...
>                     ^
> ```
> 
> Pg prefers roles that start with alpha-characters instead. I don't know if some change in this PR changed some existing behaviour (I didn't dig through), but can this be taken care of, too?

Yeah, was a new test that created a new user to ensure that would be used, will just switch it to using only random letters

akhilnarang committed:
fix(drop-site): set db root credentials in frappe.flags like how new-site is done

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: postgres uses double quotes for identifiers, especially if case-sensitive

(single quotes would force it to lowercase)

Also unconditionally create a new user while running tests - we're gonna be dropping the test site

Don't pass a `None` value to bench cli

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

akhilnarang committed:
fix: format

Got messed up during rebase - forgot to run pre-commit before `git rebase --continue`

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented on frappe/utils/backups.py:
```diff
@@ -559,7 +559,7 @@ def new_backup(
 	odb = BackupGenerator(
 		frappe.conf.db_name,
-		frappe.conf.db_name,
+		frappe.conf.db_user or frappe.conf.db_name,
```
Why is fallback required here?

akhilnarang replied:
There was one case where it wasn't getting set here

ankush commented on frappe/installer.py:
```diff
@@ -563,6 +585,11 @@ def make_site_config(
 			if db_port:
 				site_config["db_port"] = db_port
 
+			if db_user:
+				site_config["db_user"] = db_user
+			elif db_name:
+				site_config["db_user"] = db_name
```
```suggestion
			site_config["db_user"] = db_user or db_name
```

This is identical ig?

ankush commented on frappe/database/database.py:
```diff
@@ -79,7 +78,7 @@ def __init__(
 		self.host = host or frappe.conf.db_host
 		self.port = port or frappe.conf.db_port
-		self.user = user or frappe.conf.db_name
+		self.user = user or frappe.conf.db_user or frappe.conf.db_name
```
This fallback shouldn't be required?

akhilnarang replied:
There was one case where it wasn't getting set here

akhilnarang committed:
fix: drop a fallback

Had some issues earlier - seems to be resolved now

Signed-off-by: Akhil Narang <me@akhilnarang.dev>

ankush commented:
@akhilnarang can you update this in docs?

akhilnarang commented:
> @akhilnarang can you update this in docs?

Updated
- https://frappeframework.com/docs/user/en/basics/site_config
- https://frappeframework.com/docs/user/en/bench/reference/new-site
